trin-core/src/utils/provider.rs

```diff
@@ -0,0 +1,120 @@
+/// a String, which can be used to create a `Client` when necessary.
+#[derive(Clone, Default)]
+pub struct TrustedProvider {
+    pub http: Option<Request>,
+    pub ws: Option<String>,
+}
+
+impl TrustedProvider {
+    pub fn from_trin_config(trin_config: &TrinConfig) -> Self {
+        let trusted_http_client = match trin_config.trusted_provider.as_str() {
+            "infura" => Some(build_infura_http_client_from_env()),
+            "geth" => match trin_config.geth_url.clone() {
+                Some(val) => Some(build_geth_http_client_from_env(val)),
+                None => {
+                    panic!("Must supply --geth-url cli flag to use Geth as a trusted provider.")
+                }
+            },
+            _ => panic!("Invalid cli arg for --trusted-provider"),
+        };
+        let trusted_ws_url = match trin_config.bridge {
+            true => match trin_config.trusted_provider.as_str() {
+                "infura" => Some(build_infura_ws_url_from_env()),
+                "geth" => panic!("Geth connection is not currently supported over websockets."),
+                _ => panic!("Invalid cli arg for --trusted-provider"),
+            },
+            false => None,
+        };
+        Self {
+            http: trusted_http_client,
+            ws: trusted_ws_url,
+        }
+    }
+
+    pub fn dispatch_http_request(&self, method: String, params: Params) -> anyhow::Result<Value> {
+        let client = match &self.http {
+            Some(val) => val,
+            None => return Err(anyhow!("Trusted HTTP client is unavailable.")),
+        };
+        let request = JsonRequest {
+            jsonrpc: "2.0".to_string(),
+            params,
+            method,
+            id: 1,
+        };
+        match dispatch_trusted_http_request(request, client.clone()) {
+            Ok(val) => Ok(serde_json::from_str(&val)?),
+            Err(msg) => {
+                return Err(anyhow!(
+                    "Unable to request validation data from trusted provider: {:?}",
+                    msg
+                ))
+            }
+        }
+    }
+}
+
+fn get_infura_project_id_from_env() -> String {
+    match env::var("TRIN_INFURA_PROJECT_ID") {
+        Ok(val) => val,
+        Err(_) => panic!(
+            "Must supply Infura key as environment variable, like:\n\
+            TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
+        ),
+    }
+}
+
+fn build_infura_http_client_from_env() -> ureq::Request {
+    let infura_project_id = get_infura_project_id_from_env();
+    let infura_url = format!("{}{}", INFURA_BASE_HTTP_URL, infura_project_id);
+    ureq::post(&infura_url)
+}
+
+fn build_infura_ws_url_from_env() -> String {
+    let infura_project_id = get_infura_project_id_from_env();
+    format!("{}{}", INFURA_BASE_WS_URL, infura_project_id)
+}
+
+fn get_geth_client_id_from_env() -> String {
+    match env::var("GETH_CLIENT_ID") {
```

Comment:
I see we are starting to use a lot of envs, Are there any plans to introduce a configuration file in the future, something like `config.toml` for example?

Reply:
I believe there used to be an issue for it, but it might have been closed. For the time being, I'm alright with env variables - especially since I feel less likely to leak secrets by including a `.env` file in a commit than a `config.toml` file. But that's just personal preference. At some point we'll definitely want a config file